### PR TITLE
Dimensional counters

### PR DESCRIPTION
**Pull request by lucasdicioccio**

### CODE REVIEW

**Comment by lucasdicioccio** [comment]
This change breaks consumers of the API, and maybe some old GHC/Base versions. (Built with `The Glorious Glasgow Haskell Compilation System, version 8.2.2`).

I plan to wrap the new module with a type-fancy layer (either in this package or in another package) for a stronger guarantee that the dimensions passed to `lookupOrCreate` are correct (possibly saving a runtime check about dimension lengths).

Minimal compatibility patches to demonstrate the feature are provided for `ekg-json` and `ekg`.
- https://github.com/lucasdicioccio/ekg-json/tree/dimensional-counters
- https://github.com/lucasdicioccio/ekg/tree/dimensional-counters

The example works and displays individual curves flattening the dimensions as follows.
screen shot 2018-02-03 at 22 06 19

I'm more interested in providing support into `ekg-core` than `ekg`. Though, I don't mind doing a bit more. In particular, one annoying issue of my current design is the lack of "forced ordering" on the key/value pairs for the dimensions: thus preventing the `ekg` Snap layer to guess the HashMap key for retrieving a particular Sample value (but I'm not aware of users of this Snap route and I would suggest to drop the feature rather than complexifying the `ekg-core` logic for no users).

**Comment by lucasdicioccio** [comment]
@tibbe any comment about this MR?

I'm willing to provide a typeful layer above this one to enforce more type-safety between the construction and consuming sides, though it should probably go in another package where we can allow stricter GHC versions.

File: System/Metrics/Dimensional.hs [hunk 1]
```diff
@@ -0,0 +1,88 @@
...
+import Data.HashMap.Strict (HashMap, empty)
+import qualified Data.HashMap.Strict as HashMap
+
+type Name = Text
+type Explanation = Text
+type Dimensions = [Text]
+type Point = [Text]
+
+data Dimensional a = Dimensional {
+    _create      :: !(IO a)
```
Comment (tibbe) [inline System/Metrics/Dimensional.hs]: Please stick to the field naming pattern used for other data types (i.e. no leading underscore, but prefixed as needed).

File: System/Metrics/Dimensional.hs [hunk 2]
```diff
@@ -0,0 +1,88 @@
...
+    Left NotFound -> create d pt
+    Left (UnmatchedDimensions err) -> throwIO err
+    Right x -> return x
+
+type Counter = Dimensional Counter.Counter
+type Gauge = Dimensional Gauge.Gauge
+type Label = Dimensional Label.Label
+type Distribution = Dimensional Distribution.Distribution
+
+example :: IO ()
```
Comment (tibbe) [inline System/Metrics/Dimensional.hs]: This belongs in the examples/ directory and/or in some docs.

File: System/Metrics/Dimensional.hs [hunk 3]
```diff
@@ -0,0 +1,88 @@
...
+
+import Prelude hiding (lookup)
+import GHC.Exception (Exception)
+import Control.Exception (throwIO)
+import Data.Text (Text)
+import Data.IORef (IORef, atomicModifyIORef', newIORef, readIORef)
+import Data.HashMap.Strict (HashMap, empty)
+import qualified Data.HashMap.Strict as HashMap
+
+type Name = Text
```
Comment (tibbe) [inline System/Metrics/Dimensional.hs]: Lots of Haddoc docs needed here.

File: System/Metrics/Dimensional.hs [hunk 4]
```diff
@@ -0,0 +1,88 @@
+{-# LANGUAGE LambdaCase #-}
```
Comment (tibbe) [inline System/Metrics/Dimensional.hs]: This file needs some top-level Haddocks explaining how to use this module and what it's for.

**Comment by 23Skidoo** [comment]
Closing in favour of #32.

**Comment by lucasdicioccio** [comment]
Thanks @matth-

**Pull request closed without merging**